cpuset/cpuset.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// ToSliceNoSort returns a slice of integers that contains all CPU IDs from
+// this set.
+func (s CPUSet) ToSliceNoSort() []int {
+	result := make([]int, 0, len(s.cpuIDs))
+	for cpu := range s.cpuIDs {
+		result = append(result, cpu)
+	}
+	return result
+}
+
+// ToSliceInt64 returns an ordered slice of int64 that contains all CPU IDs from
+// this set
+func (s CPUSet) ToSliceInt64() []int64 {
+	result := make([]int64, 0, len(s.cpuIDs))
+	for cpu := range s.cpuIDs {
+		result = append(result, int64(cpu))
+	}
+	sort.Slice(result, func(i, j int) bool { return result[i] < result[j] })
+	return result
+}
+
+// ToSliceNoSortInt64 returns a slice of int64 that contains all CPU IDs from
+// this set.
+func (s CPUSet) ToSliceNoSortInt64() []int64 {
+	result := make([]int64, 0, len(s.cpuIDs))
+	for cpu := range s.cpuIDs {
+		result = append(result, int64(cpu))
+	}
+	return result
+}
+
+// String returns a new string representation of the CPU IDs in this CPU set
+// in canonical linux CPU list format.
+//
+// See: http://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS
+func (s CPUSet) String() string {
+	if s.IsEmpty() {
+		return ""
+	}
+
+	cpuIDs := s.ToSlice()
+
+	type rng struct {
+		start int
+		end   int
+	}
+
+	ranges := []rng{{cpuIDs[0], cpuIDs[0]}}
+
+	for i := 1; i < len(cpuIDs); i++ {
+		lastRange := &ranges[len(ranges)-1]
+		// if this CPU ID is adjacent to the high end of the last range
+		if cpuIDs[i] == lastRange.end+1 {
+			// then extend the last range to include this CPU ID
+			lastRange.end = cpuIDs[i]
+			continue
+		}
+		// otherwise, start a new range beginning with this CPU ID
+		ranges = append(ranges, rng{cpuIDs[i], cpuIDs[i]})
+	}
+
+	// construct string from ranges
+	var result bytes.Buffer
+	for _, r := range ranges {
+		if r.start == r.end {
+			result.WriteString(strconv.Itoa(r.start))
+		} else {
+			result.WriteString(fmt.Sprintf("%d-%d", r.start, r.end))
+		}
+		result.WriteString(",")
+	}
+	return strings.TrimRight(result.String(), ",")
+}
+
+// Parse CPUSet constructs a new CPU set from a Linux CPU list formatted string.
+//
+// See: http://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS
+func Parse(s string) (CPUSet, error) {
+	b := NewBuilder()
+
+	// Handle empty string.
+	if s == "" {
+		return b.Result(), nil
+	}
+
+	// Split CPU list string:
+	// "0-5,34,46-48" => ["0-5", "34", "46-48"]
+	ranges := strings.Split(s, ",")
+
+	for _, r := range ranges {
+		boundaries := strings.SplitN(r, "-", 2)
+		if len(boundaries) == 1 {
+			// Handle ranges that consist of only one CPU ID like "34".
+			cpuID, err := strconv.Atoi(boundaries[0])
+			if err != nil {
+				return NewCPUSet(), err
+			}
+			b.Add(cpuID)
+		} else if len(boundaries) == 2 {
+			// Handle multi-CPU ID ranges like "0-5".
+			start, err := strconv.Atoi(boundaries[0])
+			if err != nil {
+				return NewCPUSet(), err
+			}
+			end, err := strconv.Atoi(boundaries[1])
+			if err != nil {
+				return NewCPUSet(), err
+			}
+			if start > end {
```

Comment:
```suggestion
			if start >= end {
```

should it be `>=`? Looking at the error message...

Reply:
the reason I was looking at `Parse`/`String` methods is that the k/k code has a dependency on the checpoint file serialization format. Now the logic of serialization is moved to the external dependency and somehow we need to make sure that the serialization will keep working. Let's say we find a bug here where something like `1,2-2,3` was allowed but not any longer. Or somebody will decide to support extra space after comma. The checkpoint deserializaiton may fail. We need to make sure there is an extensive tests added on k/k side that validates that this new external dependency is not breaking back compat.

Reply:
I see comment below that `1-1` is accepteable. So it is likely just an error message problem. But nevertheless, the testing requirement stands.

@fromanirh

Reply:
@SergeyKanzhelev , what is your recommendation here? Do you want me to add more unit tests once we land in here, or in k/k? Happy to take your direction on this, but I do want to move forward in some way.

Reply:
The best path, IMO, is to fix it in k/k and then port here with no code changes

Reply:
since it's only error message fix, @thockin recommendation is the best.